lib/vfs/src/overlay_fs.rs

```diff
@@ -0,0 +1,573 @@
+use std::{fmt::Debug, path::Path};
+
+use crate::{
+    FileOpener, FileSystem, FileSystemExt, FileSystems, FsError, Metadata, OpenOptions,
+    OpenOptionsConfig, ReadDir, VirtualFile,
+};
+
+/// A primary filesystem and chain of secondary filesystems that are overlayed
+/// on top of each other.
+///
+/// # Precedence
+///
+/// The [`OverlayFileSystem`] will execute operations based on precedence.
+///
+///
+/// Most importantly, this means earlier filesystems can shadow files and
+/// directories that have a lower precedence.
+///
+///# Examples
+///
+/// Something useful to know is that the [`FileSystems`] trait is implemented
+/// for both arrays and tuples.
+///
+/// For example, if you want to create a [`crate::FileSystem`] which will
+/// create files in-memory while still being able to read from the host, you
+/// might do something like this:
+///
+/// ```rust
+/// use wasmer_vfs::{
+///     mem_fs::FileSystem as MemFS,
+///     host_fs::FileSystem as HostFS,
+///     OverlayFileSystem,
+/// };
+/// let fs = OverlayFileSystem::new(MemFS::default(), [HostFS]);
+///
+/// // This also has the benefit of storing the two values in-line with no extra
+/// // overhead or indirection.
+/// assert_eq!(
+///     std::mem::size_of_val(&fs),
+///     std::mem::size_of::<(MemFS, HostFS)>(),
+/// );
+/// ```
+///
+/// A more complex example is
+#[derive(Clone, PartialEq, Eq)]
+pub struct OverlayFileSystem<P, S> {
+    primary: P,
+    secondaries: S,
```

Comment:
I feel like this is not quite the right layer of abstraction.

I would instead have this just take a single secondary filesystem, not a list.
And in turn have another wrapper that composes multiple read-only filesystems in order.  

That also allows implementing more efficient special implementations (like for example multiple read-only WebCs) without then taking the additional overhead that always assuming a list of secondaries introduces here.

Reply:
Isn't a "wrapper that composes multiple read-only filesystems in order" almost exactly what `OverlayFileSystem` is, though?

The only difference is that this implementation would allow the first item to be mutable, but otherwise you've got a bunch of filesystems that look like they are overlayed on top of each other, using a particular order to solve precedence of operations.

Reply:
I disagree on that take @theduke. I think multiple read-only fs is the optimal approach, also because of the usability aspect. Your comment is still addressable as a subcase of OverlayFS but with only one read-only layer instead of multiple.
I'm not sure if I see the overhead that you are referring to. Can you expand on that?

This strategy of using mutliple read-only is also what other VFS implementations do:
* Rust's vfs: https://docs.rs/vfs/latest/vfs/impls/overlay/struct.OverlayFS.html
* JS memfs UnionFs: https://github.com/streamich/unionfs/blob/master/src/union.ts#L338-L349 (they are a bit more messy on their naming)

Reply:
You can always layer another more optimized abstraction into the list with a small bit of overhead, but I can see quite a few more optimized implementations where the list is mostly noise.

I think looking at things like overlayfs and the Linux kernel in general would be educatioinal.

Reply:
(in the long term, not for this PR, obviouslyl)

Reply:
To keep things lean, please lets keep using `OverlayFS`. In the future we'll make sure we are all aligned on how to abstract the fs (I'm a bit opinionated on this front as you can see, but it comes from previous experience)